src/java/davmail/exchange/ICSCalendarValidator.java

```diff
@@ -0,0 +1,149 @@
+        int nullByteCount = 0;
+        StringBuilder invalidChars = new StringBuilder();
+
+        for (char c : content.toCharArray()) {
+            if (c == NULL_BYTE) {
+                nullByteCount++;
+            } else if ((c < 32 || c == DELETE || (c >= 128 && c <= 159))) {
+                invalidChars.append(String.format("\\u%04x,", (int)c));
+            }
+        }
+
+        // Collect all found problems
+        if (nullByteCount > 0) {
+            issues.append(nullByteCount).append(" null byte(s) found");
+        }
+        if (invalidChars.length() > 0) {
+            if (issues.length() > 0) issues.append(", ");
+            issues.append("Invalid character(s): ").append(
+                    invalidChars.substring(0, invalidChars.length() - 1));
+        }
+
+        return new ValidationResult(issues.length() == 0, issues.toString());
+    }
+
+    /**
+     * Repairs an iCalendar string by removing invalid characters.
+     * Replaces multiple consecutive invalid characters with a single space.
+     * @param content The string to repair
+     * @return The repaired string
+     */
+    public static String repairICSContent(String content) {
+        if (content == null) return null;
+        String message ="ICSCalendarValidator repair characters in ICS content:";
+        
+        StringBuilder repaired = new StringBuilder();
+        boolean lastWasInvalid = false;
+
+        for (char c : content.toCharArray()) {
+            if (isValidChar(c)) {
+                repaired.append(c);
+                lastWasInvalid = false;
+            } else if (!lastWasInvalid) {
+                repaired.append(SPACE);
+                lastWasInvalid = true;
+            }
+        }
+        String fixed = repaired.toString().trim();
+        // just put output to debug logger, only if some invalid characters has been changed.
+        if (!content.equals(fixed)){ 
+            LOGGER.debug ( message + "\n[" + content + "]\n => [" + fixed + "]\n fix complete."); 
+	}
+        return fixed ;
+    }
+
+    /**
+     * Checks if a single character is valid.
+     * A character is valid if it is:
+     * - Not a control character (ASCII 0-31)
+     * - Not a delete character (ASCII 127)
+     * - Not an invalid Unicode character (128-159)
+     * @param c The character to check
+     * @return true if the character is valid
+     */
+    private static boolean isValidChar(char c) {
+        return c > 0 && !(c == DELETE || (c >= 128 && c <= 159));
+    }
+
+    /**
+     * Result structure for validation results.
+     */
+    public static class ValidationResult {
+        private final boolean isValid;
+        private final String reason;
+
+        public ValidationResult(boolean isValid, String reason) {
+            this.isValid = isValid;
+            this.reason = reason;
+        }
+
+        public boolean isValid() { return isValid; }
+        public String showReason() { return reason; }
+    }
+}
```

Comment:
This file is missing a newline at the end. While it seems unnecessary, believe it or not, the POSIX standard requires it. :smile: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

Reply:
ok , done.